CMakeLists.txt

```diff
@@ -410,17 +410,30 @@ if(Z_FEATURE_LINK_TLS)
   endif()
 
   if(MBEDTLS_FOUND)
+    if(NOT MBEDTLS_VERSION VERSION_LESS "4.0.0")
+      message(FATAL_ERROR
+        "Mbed TLS ${MBEDTLS_VERSION} is not supported. "
+        "Please install Mbed TLS 2.x or 3.x.")
+    endif()
     message(STATUS "Using Mbed TLS ${MBEDTLS_VERSION}")
-  endif()
 
-  if(PICO_STATIC)
-    target_include_directories(${Libname}_static PRIVATE ${MBEDTLS_INCLUDE_DIRS})
-  endif()
-  if(PICO_SHARED)
-    target_include_directories(${Libname}_shared PRIVATE ${MBEDTLS_INCLUDE_DIRS})
-  endif()
+    # Start from what pkg-config gave us (usually just "mbedtls")
+    set(MBEDTLS_LINK_LIBRARIES ${MBEDTLS_LIBRARIES})
+
+    # Ensure we also link the other components by *name*
+    list(APPEND MBEDTLS_LINK_LIBRARIES mbedx509 mbedcrypto)
```

Comment:
Here it may add unversioned mbedx509 and mbedcrypto. On a system with both versions 2 and 3 installed, the unversioned names resolve to the older libs, so both versions get linked and mix ABIs.